### PR TITLE
Make the Flutter Gradle Plugin apply `FlutterExtension` to (flutter) plugins

### PR DESCRIPTION
**Pull request by gmackall**
Makes the Flutter Gradle Plugin apply the `FlutterExtension` (the class that vends `flutter.minSdkVersion`, etc) to Flutter plugins.

This allows plugin authors to use `flutter.compileSdkVersion` and the like in plugin `build.gradle` files.

Doesn't use it in templates yet - holding off till we can see how this works in the plugins repo (which will sadly have to wait till this makes the next stable 😢 )

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter_tools/gradle/build.gradle.kts
Comment (gmackall) [inline packages/flutter_tools/gradle/build.gradle.kts:20]: I copied this comment style, but also open to just removing - I don't think noting that the names we choose for every plugin aren't used elsewhere is helpful.

File: packages/flutter_tools/templates/plugin/android-java.tmpl/build.gradle.tmpl
Comment (gmackall) [inline packages/flutter_tools/templates/plugin/android-java.tmpl/build.gradle.tmpl:19]: The `plugins` block can only have a couple of things before it, which includes a buildscript block but does not include these two values.

File: packages/flutter_tools/gradle/build.gradle.kts
Comment (reidbaker) [reply packages/flutter_tools/gradle/build.gradle.kts:20]: it might be more helpful to have a list of all our names that the comment references instead.

**Comment by gmackall** [comment]
cc @johnmccutchan in case you are curious about the status - this doesn't work with the old method of applying flutter plugins (the `.flutter-plugins` way [mentioned in this block](https://github.com/flutter/flutter/blob/b5030df77ccb76dcd41b7f5154390b9dc1766f9d/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy#L666)). Still thinking about if there is a way we could get it to work. Unfortunately the cleanup that the comment hoped would happen never did, and we still reference it everywhere in code that we own, e.g. packages repo: https://github.com/search?q=repo%3Aflutter%2Fpackages%20%22.flutter-plugins%22&type=code

**Comment by gmackall** [comment]
In particular, it looks like the piece that is missing from the old `settings.gradle` templates that matters here is:
```
pluginManagement {
    def flutterSdkPath = {
        def properties = new Properties()
        file("local.properties").withInputStream { properties.load(it) }
        def flutterSdkPath = properties.getProperty("flutter.sdk")
        assert flutterSdkPath != null, "flutter.sdk not set in local.properties"
        return flutterSdkPath
    }
    settings.ext.flutterSdkPath = flutterSdkPath()
    includeBuild("${settings.ext.flutterSdkPath}/packages/flutter_tools/gradle")
}
```

**Comment by johnmccutchan** [comment]
Oof, seems like we have another long cleanup in front of us.

**Comment by gmackall** [comment]
I wonder we can somehow teach even these old apps (with the old `settings.gradle`) to resolve the plugin. It seems like it should be possible, given that the entry point is the flutter tool (which then invokes gradle). Will try to find a way today

**Comment by reidbaker** [comment]
> I wonder we can somehow teach even these old apps (with the old `settings.gradle`) to resolve the plugin. It seems like it should be possible, given that the entry point is the flutter tool (which then invokes gradle, and which we have full ability to change). Will try to find a way today

we have code that force migrates gradle files on newer flutter versions. I have seen it maybe with the clean task? I wonder if you could "fix" the older apps by adding the new code.

**Comment by gmackall** [comment]
Updated the PR to take a simpler approach. 

We have a [method where we do some configuration](https://github.com/flutter/flutter/blob/69ff23b3debe9ce1e447439557c09c0bf84bb24b/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy#L762C18-L762C40) for each of the plugin projects. The new approach just applies the `FlutterExtension` as a Gradle extension to each of these subprojects.

**Review by bartekpacia — APPROVED** [review]
LGTM, good idea and I like it.

That said, I have a bad feeling that something will explode because of this.

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy
Comment (johnmccutchan) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:765]: You are dereferencing pluginProject before checking if it is null here.

Move the check on line 768 up.

Even better, you could combine them:

if (pluginProject == null || pluginProject.extensions.findByType(FlutterExtension) != null) {
  return
}

I prefer to structure code so that it returns early rather than nesting scope. I just find that control flow easier ot reason about but it isn't strictly required.

File: packages/flutter_tools/templates/plugin/android-java.tmpl/build.gradle.tmpl
Comment (johnmccutchan) [inline packages/flutter_tools/templates/plugin/android-java.tmpl/build.gradle.tmpl:31]: I think you should add a comment whenever you reference a 'flutter.variable':

// We recommend using flutter.compileSdkVersion as it will automatically
// pickup the flutter sdk's preferred version number. You can override this if
// necessary.

File: packages/flutter_tools/templates/plugin_ffi/android.tmpl/build.gradle.tmpl
Comment (johnmccutchan) [inline packages/flutter_tools/templates/plugin_ffi/android.tmpl/build.gradle.tmpl:36]: same comment about 'flutter.' variables.

File: packages/integration_test/android/build.gradle
Comment (johnmccutchan) [inline packages/integration_test/android/build.gradle:47]: here too

**Review by johnmccutchan — DISMISSED** [review]

**Comment by reidbaker** [comment]
> LGTM, good idea and I like it.
> 
> That said, I have a bad feeling that something will explode because of this.

@bartekpacia we know gradle changes especially ones that are dynamically applied are generically dangerous. If you have any suggestions for how to make this safer I/we are willing to listen and modify the code. That said, gray and I agree that this is better than breaking old app when they update a dependency that uses these new values and it is better than throwing and error and catching that error with mitigation instructions for apps. Those were 2 of the other plans. 

Do you think we should go with one of those alternatives? In my mind this is more like the flutter tool knowing that it has a dependency and adding that dependency to the tree. In theory if someone has a existing variable named flutter we could dynamically break them but that is my only forecasted break.

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy
Comment (gmackall) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:765]: Whoops, thanks for catching. I moved to after the null check. 

I can't easily combine because the portions after
```
        project.dependencies {
            api(pluginProject)
        }
       ...
```
Are independent of this check

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy
Comment (gmackall) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:766]: These are actually importing from the same class - the Flutter Gradle Plugin also [applies the `FlutterExtension` as a Gradle extension](https://github.com/flutter/flutter/blob/9af6bae6b9a8d7a8a363467b834f52bfc64c9336/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy#L203)

File: packages/flutter_tools/templates/plugin/android-java.tmpl/build.gradle.tmpl
Comment (gmackall) [reply packages/flutter_tools/templates/plugin/android-java.tmpl/build.gradle.tmpl:31]: Hmm this would mean waiting for the next stable, right? Because the packages repo tests require passing on stable and master, and this won't be available on stable

File: packages/integration_test/android/build.gradle
Comment (gmackall) [reply packages/integration_test/android/build.gradle:47]: Hmm what's the argument for adding the comment to our `integration_test` Gradle files? Just to put it in more places to increase visibility?

File: packages/flutter_tools/gradle/build.gradle.kts
Comment (gmackall) [reply packages/flutter_tools/gradle/build.gradle.kts]: Resolving because we aren't making a plugin any more

File: packages/flutter_tools/templates/plugin_ffi/android.tmpl/build.gradle.tmpl
Comment (gmackall) [reply packages/flutter_tools/templates/plugin_ffi/android.tmpl/build.gradle.tmpl:43]: Added a comment at the top of the `android` block in all templates. Also added a short version of the comment to the `integration_test` `build.gradle`

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy
Comment (reidbaker) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]: but you are casting it as the sub type (right?) If you were finding/casting it as FGP then I think flutter would be the right name.

Comment (gmackall) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]: It's the same as what the FGP is doing for the root project.

Just to cover the Gradle terminology, `flutter.blah` in the root projects build.gradle isn't referencing the FGP, but rather the `flutter` extension that the FGP creates (Gradle extension != Gradle Plugin, but rather the former is something that can be applied by the latter).

The change in the PR is to make the FGP apply this same extension to both the root project, and the plugin projects. So it is the same exact extension, which is why I think the same name makes sense

Comment (reidbaker) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]: thanks for the explanation in that case the name seems fine.

**Review by reidbaker — APPROVED** [review]

File: packages/integration_test/android/build.gradle
Comment (johnmccutchan) [reply packages/integration_test/android/build.gradle:51]: my bad, unnecessary here.

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy
Comment (johnmccutchan) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]: are they though? like wouldn't the code below already have run if the extension is registered on the project?

File: packages/flutter_tools/templates/plugin/android-java.tmpl/build.gradle.tmpl
Comment (johnmccutchan) [reply packages/flutter_tools/templates/plugin/android-java.tmpl/build.gradle.tmpl:38]: Why remove the template changes? We know they are safe and would only impact newly created projects

Comment (reidbaker) [reply packages/flutter_tools/templates/plugin/android-java.tmpl/build.gradle.tmpl:38]: Well because we are not actually sure of the consequences. We are guessing that it is safe. If we use this in our own plugins then we will have a much better chance of detecting if there are weird corner cases. 

The risk is that people make plugins that we later figure out have a corner case and we have no way to contact those plugins. If we delay updating the template the risk is that if there is no unhandled corner case then we lose easier update support for the plugins made between now and when the templates eventually get updated. 

Given that all plugins are in that state now I think that is a good tradeoff. Open to feedback that I am being too cautious.

Comment (reidbaker) [reply packages/flutter_tools/templates/plugin/android-java.tmpl/build.gradle.tmpl:38]: I would add that our guesses around gradle safety are OK and not GREAT

Comment (reidbaker) [reply packages/flutter_tools/templates/plugin/android-java.tmpl/build.gradle.tmpl:38]: Yes what I am suggesting would required us to wait to use it until the rest of this code was in stable.

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy
Comment (gmackall) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]: I guess the problem here is that I don't fully understand why some plugins already had the `FlutterExtension` created (and therefore trying again caused an error). Also, not sure this will still be a problem with the new (no new plugin) approach.

I'm going to change to remove the check to try to identify the check that fails, and fully understand.

Comment (gmackall) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]: Ok looks like we don't need the check anymore with the new approach

**Comment by andrewkolos** [comment]
Hello from tools PR triage. Is this ready for merge, or is this still waiting on review from @johnmccutchan?

**Comment by gmackall** [comment]
Updated to reflect that we plan to land this functionality without using it in the templates, and instead wait till it rolls to stable, start using it in packages, make sure nothing breaks, then move it to the templates.

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy
Comment (reidbaker) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:768]: Consider adding a comment saying what this line does for future maintainers.

**Review by reidbaker — APPROVED** [review]

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy
Comment (gmackall) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:770]: done!

**Pull request merged**
